commons/zenoh-keyexpr/src/key_expr/borrowed.rs

```diff
@@ -726,96 +727,88 @@ impl fmt::Display for keyexpr {
-enum KeyExprConstructionError {
+enum KeyExprError {
     LoneDollarStar = -1,
     SingleStarAfterDoubleStar = -2,
     DoubleStarAfterDoubleStar = -3,
     EmptyChunk = -4,
     StarsInChunk = -5,
-    DollarAfterDollarOrStar = -6,
+    DollarAfterDollar = -6,
     ContainsSharpOrQMark = -7,
     ContainsUnboundDollar = -8,
 }
 
+impl KeyExprError {
+    #[cold]
+    fn err(self, s: &str) -> ZError {
+        let error = match &self {
+            Self::LoneDollarStar => anyhow!("Invalid Key Expr `{s}`: empty chunks are forbidden, as well as leading and trailing slashes"),
+            Self::SingleStarAfterDoubleStar => anyhow!("Invalid Key Expr `{s}`: `**/*` must be replaced by `*/**` to reach canon-form"),
+            Self::DoubleStarAfterDoubleStar => anyhow!("Invalid Key Expr `{s}`: `**/**` must be replaced by `**` to reach canon-form"),
+            Self::EmptyChunk => anyhow!("Invalid Key Expr `{s}`: empty chunks are forbidden, as well as leading and trailing slashes"),
+            Self::StarsInChunk => anyhow!("Invalid Key Expr `{s}`: `*` may only be preceded by `/` or `$`"),
+            Self::DollarAfterDollar => anyhow!("Invalid Key Expr `{s}`: `$` is not allowed after `$*`"),
+            Self::ContainsSharpOrQMark => anyhow!("Invalid Key Expr `{s}`: `#` and `?` are forbidden characters"),
+            Self::ContainsUnboundDollar => anyhow!("Invalid Key Expr `{s}`: `$` is only allowed in `$*`")
+        };
+        zerror!((self) error).into()
+    }
+}
+
 impl<'a> TryFrom<&'a str> for &'a keyexpr {
     type Error = ZError;
 
     fn try_from(value: &'a str) -> Result<Self, Self::Error> {
-        let mut in_big_wild = false;
-        for chunk in value.split('/') {
-            if chunk.is_empty() {
-                bail!((KeyExprConstructionError::EmptyChunk) "Invalid Key Expr `{}`: empty chunks are forbidden, as well as leading and trailing slashes", value)
-            }
-            if chunk == "$*" {
-                bail!((KeyExprConstructionError::LoneDollarStar)
-                    "Invalid Key Expr `{}`: lone `$*`s must be replaced by `*` to reach canon-form",
-                    value
-                )
-            }
-            if in_big_wild {
-                match chunk {
-                    "**" => bail!((KeyExprConstructionError::DoubleStarAfterDoubleStar)
-                        "Invalid Key Expr `{}`: `**/**` must be replaced by `**` to reach canon-form",
-                        value
-                    ),
-                    "*" => bail!((KeyExprConstructionError::SingleStarAfterDoubleStar)
-                        "Invalid Key Expr `{}`: `**/*` must be replaced by `*/**` to reach canon-form",
-                        value
-                    ),
-                    _ => {}
+        if value.is_empty() {
+            return Err(KeyExprError::EmptyChunk.err(value));
+        }
+        let bytes = value.as_bytes();
+        let mut i = 0;
+        let mut chunk_start = 0usize.wrapping_sub(1);
```

Comment:
Why not `usize::MAX`?

Reply:
Because it's strictly the same, but I find it a lot more explicit written this way: the chunk start is one character behind the first character.

Reply:
I've changed to have `chunk_start` starting at zero (it's the character after the `/` now)